### PR TITLE
[Sharding] Temporary shards for receiving transfer

### PR DESCRIPTION
**Pull request by agourlay**
This PR adds the receiving part of a shard transfer as described in https://github.com/qdrant/qdrant/issues/789

Follow up to https://github.com/qdrant/qdrant/pull/794 but with better locking scheme introduced in https://github.com/qdrant/qdrant/pull/825

### CODE REVIEW

**Comment by agourlay** [comment]
The last commit https://github.com/qdrant/qdrant/pull/829/commits/6f08ff120febdcb0e9920596e3aba09ad790140b introduces a change in semantic for updates from peer.

Previously we would refuse update not targeting a local shard, but after this change we allow forwarding updates to remote shards to handle gracefully the eventually consistent nature if shard topology changes.

File: lib/collection/src/shard/shard_holder.rs [hunk 1]
```diff
@@ -64,6 +71,55 @@ impl ShardHolder {
...
+                error: format!(
+                    "Can't create shard {shard_id} temporary directory. Error: {}",
+                    err
+                ),
+            })?;
+        Ok(shard_path)
+    }
+
+    /// Initiate temporary shard
+    pub async fn initiate_temporary_shard(
```
Comment (generall) [inline lib/collection/src/shard/shard_holder.rs:92]: Initialization of the shard is not necessary to be under the write-lock.

In fact, disk operations involved with `LocalShard::build` and `create_temporary_shard_dir` might require quite some time. Let's split initialization into several steps: 

1. Creation of dirs and temporary shard
2. Locking and inserting shard into the holder (should be fast)

File: lib/collection/src/shard/shard_holder.rs [hunk 2]
```diff
@@ -64,6 +71,55 @@ impl ShardHolder {
...
+        &mut self,
+        collection_id: CollectionId,
+        collection_path: &Path,
+        shared_config: Arc<RwLock<CollectionConfig>>,
+        shard_id: ShardId,
+    ) -> CollectionResult<()> {
+        // check if a temporary shard is already present
+        if self.temporary_shards.contains_key(&shard_id) {
```
Comment (generall) [inline lib/collection/src/shard/shard_holder.rs:100]: Let's stick to the idempotent property - if operation performed multiple times, the result should be same as if it was performed only once. Besides, I would expect that it case of multiple attempts to transfer shard we would need to initialize it multiple times.

File: lib/collection/src/collection.rs [hunk 3]
```diff
@@ -258,17 +258,49 @@ impl Collection {
...
+                )))
+            }
+            Some(shard) => match *shard {
+                Shard::Local(_) => shard,
+                Shard::Proxy(_) => shard,
+                Shard::Remote(_) => {
+                    // check temporary shards if the target is a remote shard
+                    let temporary_shard_opt = self
+                        .shards_holder
+                        .get_temporary_shard(shard_selection)
```
Comment (generall) [inline lib/collection/src/collection.rs:293]: it looks like a double read lock to me: `get_shard` already acquired the lock for `shards_holder`

**Review by generall — CHANGES_REQUESTED** [review]
Additionally we would need to propagagte updates of collection configuration into temp shards as well.

File: lib/collection/src/collection.rs (hunk 3 above)
Comment (agourlay) [reply lib/collection/src/collection.rs:293]: Good catch :+1:

File: lib/collection/src/shard/shard_holder.rs (hunk 2 above)
Comment (agourlay) [reply lib/collection/src/shard/shard_holder.rs:100]: So we would accept any subsequent initiate requests without touching the content of the temporary shard?
What was transferred stays in place.

File: lib/collection/src/collection.rs (hunk 3 above)
Comment (agourlay) [reply lib/collection/src/collection.rs]: https://github.com/qdrant/qdrant/pull/829/commits/a54803b28fd7e2b811352ac33b91a1371c375a4d

File: lib/collection/src/shard/shard_holder.rs (hunk 2 above)
Comment (agourlay) [reply lib/collection/src/shard/shard_holder.rs:103]: I kept a log statement for observability https://github.com/qdrant/qdrant/pull/829/commits/1526c3d8d0f1014e9e9c7fafd65b4eebe6c17fc8

**Comment by agourlay** [comment]
All comments have been addressed, ready for another round of review :recycle:

File: lib/collection/src/collection.rs [hunk 4]
```diff
@@ -258,17 +258,79 @@ impl Collection {
...
-            .shards_holder
-            .local_shard_by_id(shard_selection)
-            .await?;
-        local_shard.get().update(operation.clone(), wait).await
+        let shard_holder_guard = self.shards_holder.read().await;
+        let shard_opt = shard_holder_guard.get_shard(&shard_selection);
+
+        let target_shard = match shard_opt {
+            None => {
+                return Err(CollectionError::bad_shard_selection(format!(
```
Comment (generall) [inline lib/collection/src/collection.rs:313]: Theoretically, it might be possible that we create a tmp shard for which there are no Remote shard exists just yet. e.g. in case of re-sharding. Let's handle this case right away.

File: lib/collection/src/collection.rs [hunk 5]
```diff
@@ -258,17 +258,79 @@ impl Collection {
...
+    pub async fn create_temporary_shard_dir(
+        collection_path: &Path,
+        shard_id: ShardId,
+    ) -> CollectionResult<PathBuf> {
+        let shard_path = collection_path.join(format!("{shard_id}-temp"));
+        tokio::fs::create_dir_all(&shard_path)
+            .await
+            .map_err(|err| CollectionError::ServiceError {
+                error: format!(
+                    "Can't create shard {shard_id} temporary directory. Error: {}",
```
Comment (generall) [inline lib/collection/src/collection.rs:292]: that might be no so strictly required

**Review by generall — CHANGES_REQUESTED** [review]
We would also need to handle `before_drop` for temp shards.

File: lib/collection/src/collection.rs (hunk 5 above)
Comment (agourlay) [reply lib/collection/src/collection.rs]: https://github.com/qdrant/qdrant/pull/829/commits/684731b622a27cdda0e5a6725f066f0919945553

File: lib/collection/src/collection.rs (hunk 4 above)
Comment (agourlay) [reply lib/collection/src/collection.rs]: https://github.com/qdrant/qdrant/pull/829/commits/f75252370c1dd2651a0daac66e0487fe6bcba39e

File: lib/collection/src/collection.rs [hunk 6]
```diff
@@ -258,17 +258,79 @@ impl Collection {
         todo!("Handle for sender and receiver")
     }
 
+    pub async fn initiate_temporary_shard(&self, shard_id: ShardId) -> CollectionResult<()> {
+        let shard_holder_read = self.shards_holder.read().await;
+        // validate that the shard is known as a remote shard
+        if let Some(Shard::Remote(_)) = shard_holder_read.get_shard(&shard_id) {
```
Comment (generall) [inline lib/collection/src/collection.rs:264]: Same here for support of non-existing remote shard. I suppose we need to do something like:

- Check if temp shard exists
  - if exists - drop it and re-create the directory
- Crete a new temporary shard


Do not check if remote shard exists, as it is not a strict requirement.

Comment (agourlay) [reply lib/collection/src/collection.rs]: I removed the remote shard requirement & added a bit of cleanup for existing temporary shards.
https://github.com/qdrant/qdrant/pull/829/commits/46f5b4b0d337d8b2c455936a1940c61b13bd6cc1

File: lib/collection/src/collection.rs [hunk 7]
```diff
@@ -258,17 +258,87 @@ impl Collection {
...
-        let local_shard = self
-            .shards_holder
-            .local_shard_by_id(shard_selection)
-            .await?;
-        local_shard.get().update(operation.clone(), wait).await
+        let shard_holder_guard = self.shards_holder.read().await;
+        let shard_opt = shard_holder_guard.get_shard(&shard_selection);
+
+        let target_shard = match shard_opt {
+            None => {
```
Comment (e-ivkov) [inline lib/collection/src/collection.rs:313]: Does a temporary shard have a new id? That's why we can get `None` here?

Comment (agourlay) [reply lib/collection/src/collection.rs:313]: The temporary shard should get the same id.

We have decided to remove much of the restrictions here to make the nodes more reliable to being out of sync.

Comment (e-ivkov) [reply lib/collection/src/collection.rs:313]: Ok, got it

**Comment by agourlay** [comment]
The part for promoting a temporary shard into a local shard is missing and will be added in a next PR to freeze this one :memo:

**Review by generall — APPROVED** [review]

File: tests/consensus_tests/docker-compose.yaml [hunk 8]
```diff
@@ -23,5 +23,17 @@ services:
     ports:
       - "6433:6333"
       - "6434:6334"
-    command: bash -c "sleep 20 && ./qdrant --bootstrap 'http://qdrant_node_1:6335'"
+    command: bash -c "sleep 5 && ./qdrant --bootstrap 'http://qdrant_node_1:6335' --uri 'http://qdrant_node_follower:6335'"
```
Comment (agourlay) [inline tests/consensus_tests/docker-compose.yaml:26]: Did you intend to commit this change as well? It seems unrelated.

Comment (generall) [reply tests/consensus_tests/docker-compose.yaml:26]: no, it was for the debugging of the consensus

Comment (generall) [reply tests/consensus_tests/docker-compose.yaml:26]: can we still keep it?

Comment (agourlay) [reply tests/consensus_tests/docker-compose.yaml:26]: Sure :+1:
